pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -194,11 +215,78 @@ func (p *provisioningAdmitter) validateDataDirectories(request *admission.Reques
 	if response := p.validateSystemAgentDataDirectory(oldCluster, newCluster); !response.Allowed {
 		return response
 	}
-	if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != newCluster.Spec.RKEConfig.DataDirectories.Provisioning {
+	if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != distro {
+		return admission.ResponseBadRequest("Distro data directory cannot be changed after cluster creation")
+	}
+	if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != provisioning {
 		return admission.ResponseBadRequest("Provisioning data directory cannot be changed after cluster creation")
 	}
-	if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != newCluster.Spec.RKEConfig.DataDirectories.K8sDistro {
-		return admission.ResponseBadRequest("Distro data directory cannot be changed after cluster creation")
+
+	return admission.ResponseAllowed()
+}
+
+// validateDataDirectoryFormat ensures that no data directory contains a relative path, environment variables,
+// shell expressions, or references to the current or parent directory via use of "./" and "../" respectively.
+// dir is the path of the data directory, and name corresponds to a print friendly name for this data directory.
+func validateDataDirectoryFormat(dir, name string) *admissionv1.AdmissionResponse {
+	if dir == "" {
+		return admission.ResponseAllowed()
+	}
+	if !filepath.IsAbs(dir) {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory must be an absolute path", name))
+	}
+	if strings.ContainsAny(dir, "${}()`") {
```

Comment:
This implies that these names are run through shell evaluation at some point. Are we also concerned about shell expansion on `*` or `?`?

Reply:
I don't know if the scripts are run with `noglob` or not.

If we wanted to be more comprehensive we could probably go with something like 
```golang
strings.ContainsAny(dir, "*?#~=%$|&;<>{}[]()`")
```

I looked around and I couldn't find an existing golang definition of shell metacharacters for us to use here.